packages/devextreme/js/__internal/data_controller.ts

```diff
@@ -0,0 +1,265 @@
+      pValue = propName;
+      pName = this.key() as string;
+    }
+
+    return this._dataSource.loadSingle(pName, pValue);
+  }
+
+  loadFromStore(loadOptions: unknown): Promise<unknown> {
+    return this.store().load(loadOptions);
+  }
+
+  loadNextPage(): Promise<unknown> {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-expect-error
+    // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
+    this.pageIndex(1 + this.pageIndex());
+
+    return this.load();
+  }
+
+  loadOptions(): unknown {
+    return this._dataSource.loadOptions();
+  }
+
+  userData(): unknown {
+    return this._dataSource._userData;
+  }
+
+  cancel(operationId: number): void {
+    this._dataSource.cancel(operationId);
+  }
+
+  cancelAll(): void {
+    this._dataSource.cancelAll();
+  }
+
+  filter(filter?: unknown): unknown {
+    return this._dataSource.filter(filter);
+  }
+
+  addSearchFilter(storeLoadOptions: unknown): void {
+    this._dataSource._addSearchFilter(storeLoadOptions);
+  }
+
+  group(group: unknown): unknown {
+    return this._dataSource.group(group);
+  }
+
+  paginate(): unknown {
+    return this._dataSource.paginate();
+  }
+
+  pageSize(): number {
+    return this._dataSource._pageSize;
+  }
+
+  pageIndex(pageIndex?: number): number | undefined {
+    if (pageIndex === undefined) {
+      return this._dataSource.pageIndex(undefined);
+    }
+    return this._dataSource.pageIndex(pageIndex);
+  }
+
+  resetDataSource(): void {
+    this._disposeDataSource();
+  }
+
+  resetDataSourcePageIndex(): void {
+    if (this.pageIndex()) {
+      this.pageIndex(0);
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.load();
+    }
+  }
+
+  initDataSource(dataSourceOptions: DataSourceType): void {
+    this._initDataSource(dataSourceOptions);
+  }
+
+  totalCount(): number {
+    return this._dataSource.totalCount();
+  }
+
+  isLastPage(): boolean {
+    return this._dataSource.isLastPage() || !this._dataSource._pageSize;
+  }
+
+  isLoading(): boolean {
+    return this._dataSource.isLoading();
+  }
+
+  isLoaded(): boolean {
+    return this._dataSource.isLoaded();
+  }
+
+  searchValue(value?: unknown): unknown {
+    return this._dataSource.searchValue(value);
+  }
+
+  searchOperation(operation?: string): unknown {
+    return this._dataSource.searchOperation(operation);
+  }
+
+  searchExpr(expr?: unknown): unknown {
+    return this._dataSource.searchExpr(expr);
+  }
+
+  select(...args: unknown[]): unknown {
+    return this._dataSource.select(args);
+  }
+
+  key(): string | undefined {
+    return this._dataSource?.key();
+  }
+
+  keyOf(item: unknown): string {
+    return this.store().keyOf(item);
+  }
+
+  store(): DataSourceType {
+    return this._dataSource.store();
+  }
+
+  items(): unknown[] {
+    return this._dataSource.items();
+  }
+
+  itemsToDataSource(value: unknown[], key: string): void {
+    if (!this._dataSource) {
```

Comment:
Do we really want somewhere for this function to make nothing if some dataSource is already created?
If not I prefer we name it as `updateDataSourceByItems` and start it with a disposing of an existing dataSource